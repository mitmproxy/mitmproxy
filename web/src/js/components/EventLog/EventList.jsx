import React, { Component, PropTypes } from 'react'
import ReactDOM from 'react-dom'
import shallowEqual from 'shallowequal'
import AutoScroll from '../helpers/AutoScroll'
import { calcVScroll } from '../helpers/VirtualScroll'

class EventLogList extends Component {

    static propTypes = {
        events: PropTypes.array.isRequired,
        rowHeight: PropTypes.number,
    }

    static defaultProps = {
        rowHeight: 18,
    }

    constructor(props) {
        super(props)

        this.heights = {}
        this.state = { vScroll: calcVScroll() }

        this.onViewportUpdate = this.onViewportUpdate.bind(this)
    }

    componentDidMount() {
        window.addEventListener('resize', this.onViewportUpdate)
        this.onViewportUpdate()
    }

    componentWillUnmount() {
        window.removeEventListener('resize', this.onViewportUpdate)
    }

    componentDidUpdate() {
        this.onViewportUpdate()
    }

    onViewportUpdate() {
        const viewport = ReactDOM.findDOMNode(this)

        const vScroll = calcVScroll({
            itemCount: this.props.events.length,
            rowHeight: this.props.rowHeight,
            viewportTop: viewport.scrollTop,
            viewportHeight: viewport.offsetHeight,
            itemHeights: this.props.events.map(entry => this.heights[entry.id]),
        })

        if (!shallowEqual(this.state.vScroll, vScroll)) {
            this.setState({vScroll})
        }
    }

    setHeight(id, node) {
        if (node && !this.heights[id]) {
            const height = node.offsetHeight
            if (this.heights[id] !== height) {
                this.heights[id] = height
                this.onViewportUpdate()
            }
        }
    }

    render() {
        const { vScroll } = this.state
        const { events } = this.props

        return (
            <pre onScroll={this.onViewportUpdate}>
                <div style={{ height: vScroll.paddingTop }}></div>
                {events.slice(vScroll.start, vScroll.end).map(event => (
                    <div key={event.id} ref={node => this.setHeight(event.id, node)}>
                        <LogIcon event={event}/>
                        {event.message}
                    </div>
                ))}
                <div style={{ height: vScroll.paddingBottom }}></div>
            </pre>
        )
    }
}

function LogIcon({ event }) {
    const icon = { web: 'html5', debug: 'bug' }[event.level] || 'info'
    return <i className={`fa fa-fw fa-${icon}`}></i>
}

export default AutoScroll(EventLogList)
