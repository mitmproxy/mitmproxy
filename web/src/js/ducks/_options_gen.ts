/** Auto-generated by test_app.py:test_generate_options_js */
export interface OptionsState {
    add_upstream_certs_to_client_chain: boolean
    allow_hosts: string[]
    anticache: boolean
    anticomp: boolean
    block_global: boolean
    block_list: string[]
    block_private: boolean
    body_size_limit: string | undefined
    cert_passphrase: string | undefined
    certs: string[]
    ciphers_client: string | undefined
    ciphers_server: string | undefined
    client_certs: string | undefined
    client_replay: string[]
    client_replay_concurrency: number
    command_history: boolean
    confdir: string
    connect_addr: string | undefined
    connection_strategy: string
    console_focus_follow: boolean
    content_view_lines_cutoff: number
    dns_listen_host: string
    dns_listen_port: number
    dns_mode: string
    dns_server: boolean
    export_preserve_original_ip: boolean
    http2: boolean
    http2_ping_keepalive: number
    ignore_hosts: string[]
    intercept: string | undefined
    intercept_active: boolean
    keep_host_header: boolean
    key_size: number
    listen_host: string
    listen_port: number
    map_local: string[]
    map_remote: string[]
    mode: string
    modify_body: string[]
    modify_headers: string[]
    normalize_outbound_headers: boolean
    onboarding: boolean
    onboarding_host: string
    onboarding_port: number
    proxy_debug: boolean
    proxyauth: string | undefined
    rawtcp: boolean
    readfile_filter: string | undefined
    rfile: string | undefined
    save_stream_file: string | undefined
    save_stream_filter: string | undefined
    scripts: string[]
    server: boolean
    server_replay: string[]
    server_replay_ignore_content: boolean
    server_replay_ignore_host: boolean
    server_replay_ignore_params: string[]
    server_replay_ignore_payload_params: string[]
    server_replay_ignore_port: boolean
    server_replay_kill_extra: boolean
    server_replay_nopop: boolean
    server_replay_refresh: boolean
    server_replay_use_headers: string[]
    showhost: boolean
    ssl_insecure: boolean
    ssl_verify_upstream_trusted_ca: string | undefined
    ssl_verify_upstream_trusted_confdir: string | undefined
    stickyauth: string | undefined
    stickycookie: string | undefined
    stream_large_bodies: string | undefined
    tcp_hosts: string[]
    termlog_verbosity: string
    tls_version_client_max: string
    tls_version_client_min: string
    tls_version_server_max: string
    tls_version_server_min: string
    upstream_auth: string | undefined
    upstream_cert: boolean
    validate_inbound_headers: boolean
    view_filter: string | undefined
    view_order: string
    view_order_reversed: boolean
    web_columns: string[]
    web_debug: boolean
    web_host: string
    web_open_browser: boolean
    web_port: number
    web_static_viewer: string | undefined
    websocket: boolean
}

export type Option = keyof OptionsState

export const defaultState: OptionsState = {
    add_upstream_certs_to_client_chain: false,
    allow_hosts: [],
    anticache: false,
    anticomp: false,
    block_global: true,
    block_list: [],
    block_private: false,
    body_size_limit: undefined,
    cert_passphrase: undefined,
    certs: [],
    ciphers_client: undefined,
    ciphers_server: undefined,
    client_certs: undefined,
    client_replay: [],
    client_replay_concurrency: 1,
    command_history: true,
    confdir: "~/.mitmproxy",
    connect_addr: undefined,
    connection_strategy: "eager",
    console_focus_follow: false,
    content_view_lines_cutoff: 512,
    dns_listen_host: "",
    dns_listen_port: 53,
    dns_mode: "regular",
    dns_server: false,
    export_preserve_original_ip: false,
    http2: true,
    http2_ping_keepalive: 58,
    ignore_hosts: [],
    intercept: undefined,
    intercept_active: false,
    keep_host_header: false,
    key_size: 2048,
    listen_host: "",
    listen_port: 8080,
    map_local: [],
    map_remote: [],
    mode: "regular",
    modify_body: [],
    modify_headers: [],
    normalize_outbound_headers: true,
    onboarding: true,
    onboarding_host: "mitm.it",
    onboarding_port: 80,
    proxy_debug: false,
    proxyauth: undefined,
    rawtcp: true,
    readfile_filter: undefined,
    rfile: undefined,
    save_stream_file: undefined,
    save_stream_filter: undefined,
    scripts: [],
    server: true,
    server_replay: [],
    server_replay_ignore_content: false,
    server_replay_ignore_host: false,
    server_replay_ignore_params: [],
    server_replay_ignore_payload_params: [],
    server_replay_ignore_port: false,
    server_replay_kill_extra: false,
    server_replay_nopop: false,
    server_replay_refresh: true,
    server_replay_use_headers: [],
    showhost: false,
    ssl_insecure: false,
    ssl_verify_upstream_trusted_ca: undefined,
    ssl_verify_upstream_trusted_confdir: undefined,
    stickyauth: undefined,
    stickycookie: undefined,
    stream_large_bodies: undefined,
    tcp_hosts: [],
    termlog_verbosity: "info",
    tls_version_client_max: "UNBOUNDED",
    tls_version_client_min: "TLS1_2",
    tls_version_server_max: "UNBOUNDED",
    tls_version_server_min: "TLS1_2",
    upstream_auth: undefined,
    upstream_cert: true,
    validate_inbound_headers: true,
    view_filter: undefined,
    view_order: "time",
    view_order_reversed: false,
    web_columns: ["tls", "icon", "path", "method", "status", "size", "time"],
    web_debug: false,
    web_host: "127.0.0.1",
    web_open_browser: true,
    web_port: 8081,
    web_static_viewer: "",
    websocket: true,
}
