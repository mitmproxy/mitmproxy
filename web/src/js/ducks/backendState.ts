/**
 * serverState houses properties about the current mitmproxy instance that are not options,
 * e.g. the list of available content views or the current version.
 */

export const RECEIVE = 'STATE_RECEIVE'
export const UPDATE = 'STATE_UPDATE'

export interface ServerInfo {
    description: string
    full_spec: string
    is_running: boolean
    last_exception: string | null
    listen_addrs: [string, number][]
    type: string
    wireguard_conf?: string
}

export interface BackendState {
    available: boolean
    version: string
    contentViews: string[]
    servers: ServerInfo[]
}

export const defaultState: BackendState = {
    available: false,
    version: "",
    contentViews: [],
    servers: [],
};

export default function reducer(state = defaultState, action): BackendState {
    switch (action.type) {
        case RECEIVE:
        case UPDATE:
            return {
                ...state,
                available: true,
                ...action.data
            }
        default:
            return state
    }
}
