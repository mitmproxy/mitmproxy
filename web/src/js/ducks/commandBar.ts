export const TOGGLE_VISIBILITY = "COMMANDBAR_TOGGLE_VISIBILITY";

interface CommandBarState {
    visible: boolean;
}

export const defaultState: CommandBarState = {
    visible: false,
};

export default function reducer(state = defaultState, action): CommandBarState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        default:
            return state;
    }
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}
