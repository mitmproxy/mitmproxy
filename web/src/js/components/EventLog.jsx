import React, { PropTypes } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { toggleEventLogFilter, toggleEventLogVisibility } from '../ducks/eventLog'
import { ToggleButton } from './common'
import EventList from './EventLog/EventList'

EventLog.propTypes = {
    filters: PropTypes.object.isRequired,
    events: PropTypes.array.isRequired,
    onToggleFilter: PropTypes.func.isRequired,
    onClose: PropTypes.func.isRequired
}

function EventLog({ filters, events, onToggleFilter, onClose }) {
    return (
        <div className="eventlog">
            <div>
                Eventlog
                <div className="pull-right">
                    {['debug', 'info', 'web'].map(type => (
                        <ToggleButton key={type} text={type} checked={filters[type]} onToggle={() => onToggleFilter(type)}/>
                    ))}
                    <i onClick={onClose} className="fa fa-close"></i>
                </div>
            </div>
            <EventList events={events} />
        </div>
    )
}

export default connect(
    state => ({
        filters: state.eventLog.filter,
        events: state.eventLog.filteredEvents,
    }),
    dispatch => bindActionCreators({
        onClose: toggleEventLogVisibility,
        onToggleFilter: toggleEventLogFilter,
    }, dispatch)
)(EventLog)
