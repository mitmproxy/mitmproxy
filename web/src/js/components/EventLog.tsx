import React, { Component } from "react";
import { connect } from "react-redux";
import {
    EventLogItem,
    LogLevel,
    toggleFilter,
    toggleVisibility,
} from "../ducks/eventLog";
import ToggleButton from "./common/ToggleButton";
import EventList from "./EventLog/EventList";
import { RootState } from "../ducks";

type EventLogState = {
    height: number;
};

type EventLogProps = {
    events: EventLogItem[];
    filters: { [level in LogLevel]: boolean };
    toggleFilter: (filter: LogLevel) => any;
    close: () => any;
    defaultHeight: number;
};

export class PureEventLog extends Component<EventLogProps, EventLogState> {
    static defaultProps = {
        defaultHeight: 200,
    };
    private dragStart: number;

    constructor(props, context) {
        super(props, context);

        this.state = { height: this.props.defaultHeight };

        this.onDragStart = this.onDragStart.bind(this);
        this.onDragMove = this.onDragMove.bind(this);
        this.onDragStop = this.onDragStop.bind(this);
    }

    onDragStart(event: React.MouseEvent) {
        event.preventDefault();
        this.dragStart = this.state.height + event.pageY;
        window.addEventListener("mousemove", this.onDragMove);
        window.addEventListener("mouseup", this.onDragStop);
        window.addEventListener("dragend", this.onDragStop);
    }

    onDragMove(event: MouseEvent) {
        event.preventDefault();
        this.setState({ height: this.dragStart - event.pageY });
    }

    onDragStop(event: MouseEvent | DragEvent) {
        event.preventDefault();
        window.removeEventListener("mousemove", this.onDragMove);
    }

    render() {
        const { height } = this.state;
        const { filters, events, toggleFilter, close } = this.props;

        return (
            <div className="eventlog" style={{ height }}>
                <div onMouseDown={this.onDragStart}>
                    Eventlog
                    <div className="pull-right">
                        {Object.values(LogLevel).map((type) => (
                            <ToggleButton
                                key={type}
                                text={type}
                                checked={filters[type]}
                                onToggle={() => toggleFilter(type)}
                            />
                        ))}
                        <i onClick={close} className="fa fa-close"></i>
                    </div>
                </div>
                <EventList events={events} />
            </div>
        );
    }
}

export default connect(
    (state: RootState) => ({
        filters: state.eventLog.filters,
        events: state.eventLog.view,
    }),
    {
        close: toggleVisibility,
        toggleFilter: toggleFilter,
    },
)(PureEventLog);
