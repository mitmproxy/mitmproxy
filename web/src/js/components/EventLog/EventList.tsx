import React, { Component } from "react";
import PropTypes from "prop-types";
import ReactDOM from "react-dom";
import shallowEqual from "shallowequal";
import AutoScroll from "../helpers/AutoScroll";
import { calcVScroll, VScroll } from "../helpers/VirtualScroll";
import { EventLogItem } from "../../ducks/eventLog";

type EventLogListProps = {
    events: EventLogItem[];
    rowHeight: number;
};
type EventLogListState = {
    vScroll: VScroll;
};

class EventLogList extends Component<EventLogListProps, EventLogListState> {
    static propTypes = {
        events: PropTypes.array.isRequired,
        rowHeight: PropTypes.number,
    };

    static defaultProps = {
        rowHeight: 18,
    };

    heights: { [id: string]: number };

    constructor(props) {
        super(props);

        this.heights = {};
        this.state = { vScroll: calcVScroll() };

        this.onViewportUpdate = this.onViewportUpdate.bind(this);
    }

    componentDidMount() {
        window.addEventListener("resize", this.onViewportUpdate);
        this.onViewportUpdate();
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this.onViewportUpdate);
    }

    componentDidUpdate() {
        this.onViewportUpdate();
    }

    onViewportUpdate() {
        const viewport = ReactDOM.findDOMNode(this);

        const vScroll = calcVScroll({
            itemCount: this.props.events.length,
            rowHeight: this.props.rowHeight,
            viewportTop: viewport.scrollTop,
            viewportHeight: viewport.offsetHeight,
            itemHeights: this.props.events.map(
                (entry) => this.heights[entry.id]
            ),
        });

        if (!shallowEqual(this.state.vScroll, vScroll)) {
            this.setState({ vScroll });
        }
    }

    setHeight(id, node) {
        if (node && !this.heights[id]) {
            const height = node.offsetHeight;
            if (this.heights[id] !== height) {
                this.heights[id] = height;
                this.onViewportUpdate();
            }
        }
    }

    render() {
        const { vScroll } = this.state;
        const { events } = this.props;

        return (
            <pre onScroll={this.onViewportUpdate}>
                <div style={{ height: vScroll.paddingTop }} />
                {events.slice(vScroll.start, vScroll.end).map((event) => (
                    <div
                        key={event.id}
                        ref={(node) => this.setHeight(event.id, node)}
                    >
                        <LogIcon event={event} />
                        {event.message}
                    </div>
                ))}
                <div style={{ height: vScroll.paddingBottom }} />
            </pre>
        );
    }
}

function LogIcon({ event }) {
    const icon =
        {
            web: "html5",
            debug: "bug",
            warn: "exclamation-triangle",
            error: "ban",
        }[event.level] || "info";
    return <i className={`fa fa-fw fa-${icon}`} />;
}

export default AutoScroll(EventLogList);
