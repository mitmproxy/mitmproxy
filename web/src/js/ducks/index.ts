import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
