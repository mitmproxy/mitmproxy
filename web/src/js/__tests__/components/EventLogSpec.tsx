jest.mock("../../components/EventLog/EventList");

import * as React from "react";
import renderer from "react-test-renderer";
import EventLog, { PureEventLog } from "../../components/EventLog";
import { Provider } from "react-redux";
import { TStore } from "../ducks/tutils";

window.addEventListener = jest.fn();
window.removeEventListener = jest.fn();

describe("EventLog Component", () => {
    let store = TStore(),
        provider = renderer.create(
            <Provider store={store}>
                <EventLog />
            </Provider>,
        ),
        tree = provider.toJSON();

    it("should connect to state and render correctly", () => {
        expect(tree).toMatchSnapshot();
    });

    it("should handle toggleFilter", () => {
        const debugToggleButton = tree.children[0].children[1].children[0];
        debugToggleButton.props.onClick();
    });

    provider = renderer.create(
        <Provider store={store}>
            <EventLog />
        </Provider>,
    );
    const eventLog = provider.root.findByType(PureEventLog);
    const mockEvent = { preventDefault: jest.fn() };

    it("should handle DragStart", () => {
        eventLog.instance.onDragStart(mockEvent);
        expect(mockEvent.preventDefault).toBeCalled();
        expect(window.addEventListener).toBeCalledWith(
            "mousemove",
            eventLog.instance.onDragMove,
        );
        expect(window.addEventListener).toBeCalledWith(
            "mouseup",
            eventLog.instance.onDragStop,
        );
        expect(window.addEventListener).toBeCalledWith(
            "dragend",
            eventLog.instance.onDragStop,
        );
        mockEvent.preventDefault.mockClear();
    });

    it("should handle DragMove", () => {
        eventLog.instance.onDragMove(mockEvent);
        expect(mockEvent.preventDefault).toBeCalled();
        mockEvent.preventDefault.mockClear();
    });

    console.error = jest.fn(); // silent the error.
    it("should handle DragStop", () => {
        eventLog.instance.onDragStop(mockEvent);
        expect(mockEvent.preventDefault).toBeCalled();
        expect(window.removeEventListener).toBeCalledWith(
            "mousemove",
            eventLog.instance.onDragMove,
        );
    });
});
