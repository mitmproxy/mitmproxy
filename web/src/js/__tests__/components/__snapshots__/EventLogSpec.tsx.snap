// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventLog Component should connect to state and render correctly 1`] = `
<div
  className="eventlog"
  style={
    {
      "height": 200,
    }
  }
>
  <div
    onMouseDown={[Function]}
  >
    Eventlog
    <div
      className="pull-right"
    >
      <div
        className="btn btn-toggle btn-primary"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-check-square-o"
        />
         
        debug
      </div>
      <div
        className="btn btn-toggle btn-primary"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-check-square-o"
        />
         
        info
      </div>
      <div
        className="btn btn-toggle btn-default"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-square-o"
        />
         
        web
      </div>
      <div
        className="btn btn-toggle btn-primary"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-check-square-o"
        />
         
        warn
      </div>
      <div
        className="btn btn-toggle btn-primary"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-check-square-o"
        />
         
        error
      </div>
      <i
        className="fa fa-close"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
